lib/parsers/heif_parser.rb

```diff
@@ -0,0 +1,430 @@
+    end
+  end
+
+  def scan_meta_level_box
+    metadata_length = read_int_32
+    return unless read_string(4) == META_BOX_MARKER
+    @metadata_start_pos = @buf.pos
+    @metadata_end_pos = @buf.pos + metadata_length - HEADER_LENGTH # the real data is always without the 8 initial bytes of the handler
+    read_nil_version_and_flag
+
+    # we are looking for box/containers right beneath the Meta box
+    # we start with the HDLR (Handler) box..
+    handler_length = read_int_32
+    return unless read_string(4) == HANDLER_MARKER
+    handler_length -= HEADER_LENGTH # subtract the header as usual (will not be mentioned anymore from now on)
+    handler_start = @buf.pos
+    # the handler type declares the type of metadata and thus the process by which the media-data in the track is presented
+    # it also indicates the structure or format of the ‘meta’ box contents
+    read_nil_version_and_flag
+    read_string(4) # pre_defined bytes, always 4 null bytes in the hdlr box
+    @handler_type = read_string(4)
+    @buf.seek(handler_start + handler_length) # the remaining part is reserved
+
+    # ..continue looking for the IINF box and especially for the IPRP box, containing info about the image itself
+    next_box_length = read_int_32
+    next_box = read_string(4)
+    next_box_start_pos = @buf.pos
+    while @buf.pos < @metadata_end_pos # we iterate over all next incoming boxed but without going outside the meta-box
+      case next_box
+      when PRIMARY_ITEM_BOX
+        read_primary_item_box
+      when ITEM_INFO_BOX
+        read_item_info_box
+      when ITEM_PROPERTIES_BOX
+        read_item_properties_box
+        fill_primary_values
+      when next_box == ''
+        break
+      end
+      next_box_length, next_box, next_box_start_pos = get_next_box(next_box_start_pos, next_box_length, @metadata_end_pos)
+    end
+  end
+
+  def read_item_info_box
+    version = read_int_8
+    safe_skip(@buf, 3) # 0 flags
+    entry_count = if version == 0
+      read_int_16
+    else
+      read_int_32
+    end
+    @sub_items = []
+    entry_count.times {
+      item_info_entry_length = read_int_32
+      return unless read_string(4) == ITEM_INFO_ENTRY
+      item_info_end_pos = @buf.pos + item_info_entry_length - HEADER_LENGTH
+      version = read_int_8
+      safe_skip(@buf, 3) # 0 flags
+      case version
+      when 2
+        item_id = read_int_16
+      when 3
+        item_id = read_int_32
+      else
+        return # wrong version according to standards, hence return
+      end
+      safe_skip(@buf, 2) # not interested in the item_protection_index
+      item_type = read_string(4)
+      content_encoding = ''
+      if item_type == MIME_MARKER
+        content_encoding = read_string(item_info_end_pos - @buf.pos).delete!("\0") # remove the null-termination part for output visualization reason
+      end
+      @sub_items << {'item_id': item_id, 'item_type': item_type, 'content_encoding': content_encoding}
```

Comment:
Do you wish these to be symbol keys or string keys? Because you may drop the quotes if you need symbols, or use `=>` if you need strings 😊

Reply:
good tip! Thank you! :)